src/main/java/io/anserini/collection/WikipediaParagraphsCollection.java

```diff
@@ -0,0 +1,133 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.anserini.collection;
+
+import io.anserini.document.WikipediaParagraph;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import org.wikiclean.WikiClean;
+import org.wikiclean.WikiClean.WikiLanguage;
+import org.wikiclean.WikiCleanBuilder;
+import org.wikiclean.WikipediaBz2DumpInputStream;
+
+/**
+ * Class representing an instance of a Wikipedia paragraphs collection. Note that Wikipedia dumps often come
+ * as a single bz2 file. Since a collection is assumed to be in a directory, place the bz2 file in
+ * a directory prior to indexing.
+ */
+public class WikipediaParagraphsCollection extends Collection<WikipediaParagraph> {
+
+  public class FileSegment extends Collection<WikipediaParagraph>.FileSegment {
+    private final WikipediaBz2DumpInputStream stream;
+    private final WikiClean cleaner;
+    private final Pattern CONTAINS_PERIOD;
+    private final Pattern SINGLE_WORD_WITH_PERIOD;
+    private ArrayList<String> bufferedParagraphs = null;
+    private int currentParagraphIdx = 0;
+    private String title = "";
+
+    protected FileSegment(Path path) throws IOException {
+      this.path = path;
+
+      stream = new WikipediaBz2DumpInputStream(path.toString());
+      cleaner = new WikiCleanBuilder()
+          .withLanguage(WikiLanguage.EN).withTitle(false)
+          .withFooter(false).build();
+
+      CONTAINS_PERIOD = Pattern.compile("\\.");
+      SINGLE_WORD_WITH_PERIOD = Pattern.compile("\\w+\\.");
+    }
+
+    @Override
+    public WikipediaParagraph next() {
+      if (bufferedParagraphs != null && currentParagraphIdx < bufferedParagraphs.size()) {
+        String currentParagraph = bufferedParagraphs.get(currentParagraphIdx);
+        WikipediaParagraph result = new WikipediaParagraph(title, currentParagraphIdx, currentParagraph);
+        currentParagraphIdx++;
+        return result;
+      }
+
+      try {
+        String page;
+        String s;
+
+        // Advance to the next valid page.
+        while ((page = stream.readNext()) != null) {
+          // See https://en.wikipedia.org/wiki/Wikipedia:Namespace
+          if (page.contains("<ns>") && !page.contains("<ns>0</ns>")) {
+            continue;
+          }
+
+          s = cleaner.clean(page);
+
+          // Skip redirects
+          if (s.startsWith("#REDIRECT")) {
+            continue;
+          }
+
+          String[] parts = s.split("\\n+");
+          bufferedParagraphs = new ArrayList<>();
+          for (String p : parts) {
+            // Heuristic is that parts without periods ('.') or has one word followed by a period is a section header and should not be indexed.
```

Comment:
I think it is better to explicitly add comments about how a paragraph is identified?
My understanding is that, first a page is split into lines; for each line, it has to contain at least a dot `.` in order to be included in `bufferedParagraphs`. That means a line w/o dot won't be included in any paragraphs? This seems to be counter-intuitive...